### PR TITLE
Bulk fetch of batches in executor

### PR DESCRIPTION
**Pull request by arun-koshy**
## Description 

TLDR: This PR introduces bulk fetch for all batches required to execute a committed subdag which improves catchup 2.5x

Was waiting to submit this PR in light of another PR that is being tested/experimented to introduce batching for payload fetch in the synchronizer and unite the following three batch fetch paths through a shared `BatchFetcher` worker component.

1. Synchronize header batch (blocking)
2. Synchronize certificate batches (non-blocking)
3. Synchronize certificate batches & fetch batches for execution (blocking)

The main goal of synchronizer being to get the payload into the workers local store so that at the time of fetching we are hitting our local worker and not a remote worker most of the time. The synchronizer will feed digests to the BatchFetcher which will have a queue of missing digests it needs to fetch and it can then bulk fetch those from local store or remote workers in the background or blocking if required immediately. Doing this will allow us to reduce the total number of fetch requests significantly because we are bulk fetching and deduping digests that are being fetched.

Unfortunately we have not seen major results yet with those changes as there seem to be other bottlenecks that need to be fixed first which is why we are deploying this incremental change first to speed up catchup and tail latencies. Will send the follow up PR when the experimental impact matches the risk of the refactor introduced. 

## Test Plan 

Unit tests and benchmark cluster

- 8-node geo distributed
-- **Before batching** ~ 16-72 certs per second
Example total catchup time - [2 hours of down time, about 1 hour to catchup](https://mysten.grafana.net/explore?left=%7B%22datasource%22:%228Xt1pVoVk%22,%22queries%22:%5B%7B%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%228Xt1pVoVk%22%7D,%22editorMode%22:%22builder%22,%22expr%22:%22sum%20by%28host%29%20%28last_committed_round%7Bhost%3D~%5C%22ams-bnc-val-00%7Cewr-bnc-val-00%5C%22,%20network%3D%5C%22benchmark%5C%22%7D%29%22,%22hide%22:false,%22range%22:true,%22refId%22:%22C%22,%22interval%22:%22%22%7D,%7B%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%228Xt1pVoVk%22%7D,%22editorMode%22:%22builder%22,%22expr%22:%22sum%20by%28host%29%20%28rate%28subscriber_processed_batches%7Bhost%3D~%5C%22ams-bnc-val-00%7Cewr-bnc-val-00%5C%22,%20network%3D%5C%22benchmark%5C%22%7D%5B5m%5D%29%29%22,%22legendFormat%22:%22__auto%22,%22range%22:true,%22instant%22:true,%22hide%22:true%7D%5D,%22range%22:%7B%22from%22:%221678411411288%22,%22to%22:%221678424058450%22%7D%7D&orgId=1&right=%7B%22datasource%22:%228Xt1pVoVk%22,%22queries%22:%5B%7B%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%228Xt1pVoVk%22%7D,%22editorMode%22:%22builder%22,%22expr%22:%22sum%20by%28host%29%20%28last_committed_round%7Bhost%3D~%5C%22ams-bnc-val-00%7Cewr-bnc-val-00%5C%22,%20network%3D%5C%22benchmark%5C%22%7D%29%22,%22hide%22:true,%22range%22:true,%22refId%22:%22C%22,%22interval%22:%22%22%7D,%7B%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%228Xt1pVoVk%22%7D,%22editorMode%22:%22builder%22,%22expr%22:%22sum%20by%28host%29%20%28rate%28subscriber_processed_batches%7Bhost%3D~%5C%22ams-bnc-val-00%7Cewr-bnc-val-00%5C%22,%20network%3D%5C%22benchmark%5C%22%7D%5B5m%5D%29%29%22,%22legendFormat%22:%22__auto%22,%22range%22:true,%22instant%22:true,%22hide%22:true%7D,%7B%22refId%22:%22B%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%228Xt1pVoVk%22%7D,%22editorMode%22:%22code%22,%22expr%22:%22sum%20by%20%28host%29%20%28rate%28sequencing_certificate_attempt%7Bhost%3D~%5C%22.%2A%5C%22,%20network%3D%5C%22benchmark%5C%22%7D%5B$__rate_interval%5D%29%29%22,%22legendFormat%22:%22__auto%22,%22range%22:true,%22instant%22:true%7D%5D,%22range%22:%7B%22from%22:%221678411411288%22,%22to%22:%221678424058450%22%7D%7D)
Commit round rate - [2-9 rounds/s with one spike up to 15 r/s](https://mysten.grafana.net/explore?left=%7B%22datasource%22:%228Xt1pVoVk%22,%22queries%22:%5B%7B%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%228Xt1pVoVk%22%7D,%22editorMode%22:%22builder%22,%22expr%22:%22sum%20by%28host%29%20%28rate%28last_committed_round%7Bhost%3D~%5C%22ams-bnc-val-00%7Cewr-bnc-val-00%5C%22,%20network%3D%5C%22benchmark%5C%22%7D%5B5m%5D%29%29%22,%22hide%22:false,%22range%22:true,%22refId%22:%22C%22,%22interval%22:%22%22%7D,%7B%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%228Xt1pVoVk%22%7D,%22editorMode%22:%22builder%22,%22expr%22:%22sum%20by%28host%29%20%28rate%28subscriber_processed_batches%7Bhost%3D~%5C%22ams-bnc-val-00%7Cewr-bnc-val-00%5C%22,%20network%3D%5C%22benchmark%5C%22%7D%5B5m%5D%29%29%22,%22legendFormat%22:%22__auto%22,%22range%22:true,%22instant%22:true,%22hide%22:true%7D%5D,%22range%22:%7B%22from%22:%221678394911392%22,%22to%22:%221678483239964%22%7D%7D&orgId=1&right=%7B%22datasource%22:%228Xt1pVoVk%22,%22queries%22:%5B%7B%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%228Xt1pVoVk%22%7D,%22editorMode%22:%22builder%22,%22expr%22:%22sum%20by%28host%29%20%28last_committed_round%7Bhost%3D~%5C%22ams-bnc-val-00%7Cewr-bnc-val-00%5C%22,%20network%3D%5C%22benchmark%5C%22%7D%29%22,%22hide%22:true,%22range%22:true,%22refId%22:%22C%22,%22interval%22:%22%22%7D,%7B%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%228Xt1pVoVk%22%7D,%22editorMode%22:%22builder%22,%22expr%22:%22sum%20by%28host%29%20%28rate%28subscriber_processed_batches%7Bhost%3D~%5C%22ams-bnc-val-00%7Cewr-bnc-val-00%5C%22,%20network%3D%5C%22benchmark%5C%22%7D%5B5m%5D%29%29%22,%22legendFormat%22:%22__auto%22,%22range%22:true,%22instant%22:true,%22hide%22:true%7D,%7B%22refId%22:%22B%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%228Xt1pVoVk%22%7D,%22editorMode%22:%22code%22,%22expr%22:%22sum%20by%20%28host%29%20%28rate%28sequencing_certificate_attempt%7Bhost%3D~%5C%22.%2A%5C%22,%20network%3D%5C%22benchmark%5C%22%7D%5B$__rate_interval%5D%29%29%22,%22legendFormat%22:%22__auto%22,%22range%22:true,%22instant%22:true%7D%5D,%22range%22:%7B%22from%22:%221678394911392%22,%22to%22:%221678483239964%22%7D%7D)
-- **After batching**  ~ 96-184 certs per second
Example total catchup time - [7.5 hours of down time, about 45 minutes to catchup](https://mysten.grafana.net/explore?left=%7B%22datasource%22:%228Xt1pVoVk%22,%22queries%22:%5B%7B%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%228Xt1pVoVk%22%7D,%22editorMode%22:%22builder%22,%22expr%22:%22sum%20by%28host%29%20%28last_committed_round%7Bhost%3D~%5C%22ams-bnc-val-00%7Cewr-bnc-val-00%5C%22,%20network%3D%5C%22benchmark%5C%22%7D%29%22,%22hide%22:false,%22range%22:true,%22refId%22:%22C%22,%22interval%22:%22%22%7D,%7B%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%228Xt1pVoVk%22%7D,%22editorMode%22:%22builder%22,%22expr%22:%22sum%20by%28host%29%20%28rate%28subscriber_processed_batches%7Bhost%3D~%5C%22ams-bnc-val-00%7Cewr-bnc-val-00%5C%22,%20network%3D%5C%22benchmark%5C%22%7D%5B5m%5D%29%29%22,%22legendFormat%22:%22__auto%22,%22range%22:true,%22instant%22:true,%22hide%22:true%7D%5D,%22range%22:%7B%22from%22:%221678520418067%22,%22to%22:%221678552191281%22%7D%7D&orgId=1&right=%7B%22datasource%22:%228Xt1pVoVk%22,%22queries%22:%5B%7B%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%228Xt1pVoVk%22%7D,%22editorMode%22:%22builder%22,%22expr%22:%22sum%20by%28host%29%20%28last_committed_round%7Bhost%3D~%5C%22ams-bnc-val-00%7Cewr-bnc-val-00%5C%22,%20network%3D%5C%22benchmark%5C%22%7D%29%22,%22hide%22:true,%22range%22:true,%22refId%22:%22C%22,%22interval%22:%22%22%7D,%7B%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%228Xt1pVoVk%22%7D,%22editorMode%22:%22builder%22,%22expr%22:%22sum%20by%28host%29%20%28rate%28subscriber_processed_batches%7Bhost%3D~%5C%22ams-bnc-val-00%7Cewr-bnc-val-00%5C%22,%20network%3D%5C%22benchmark%5C%22%7D%5B5m%5D%29%29%22,%22legendFormat%22:%22__auto%22,%22range%22:true,%22instant%22:true,%22hide%22:true%7D,%7B%22refId%22:%22B%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%228Xt1pVoVk%22%7D,%22editorMode%22:%22code%22,%22expr%22:%22sum%20by%20%28host%29%20%28rate%28sequencing_certificate_attempt%7Bhost%3D~%5C%22.%2A%5C%22,%20network%3D%5C%22benchmark%5C%22%7D%5B$__rate_interval%5D%29%29%22,%22legendFormat%22:%22__auto%22,%22range%22:true,%22instant%22:true%7D%5D,%22range%22:%7B%22from%22:%221678520418067%22,%22to%22:%221678552191281%22%7D%7D)
Commit round rate - [12-23 rounds/s with one spike of about 40 r/s](https://mysten.grafana.net/explore?left=%7B%22datasource%22:%228Xt1pVoVk%22,%22queries%22:%5B%7B%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%228Xt1pVoVk%22%7D,%22editorMode%22:%22builder%22,%22expr%22:%22sum%20by%28host%29%20%28rate%28last_committed_round%7Bhost%3D~%5C%22ams-bnc-val-00%7Cewr-bnc-val-00%5C%22,%20network%3D%5C%22benchmark%5C%22%7D%5B5m%5D%29%29%22,%22hide%22:false,%22range%22:true,%22refId%22:%22C%22,%22interval%22:%22%22%7D,%7B%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%228Xt1pVoVk%22%7D,%22editorMode%22:%22builder%22,%22expr%22:%22sum%20by%28host%29%20%28rate%28subscriber_processed_batches%7Bhost%3D~%5C%22ams-bnc-val-00%7Cewr-bnc-val-00%5C%22,%20network%3D%5C%22benchmark%5C%22%7D%5B5m%5D%29%29%22,%22legendFormat%22:%22__auto%22,%22range%22:true,%22instant%22:true,%22hide%22:true%7D%5D,%22range%22:%7B%22from%22:%221678520418067%22,%22to%22:%221678552191281%22%7D%7D&orgId=1&right=%7B%22datasource%22:%228Xt1pVoVk%22,%22queries%22:%5B%7B%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%228Xt1pVoVk%22%7D,%22editorMode%22:%22builder%22,%22expr%22:%22sum%20by%28host%29%20%28last_committed_round%7Bhost%3D~%5C%22ams-bnc-val-00%7Cewr-bnc-val-00%5C%22,%20network%3D%5C%22benchmark%5C%22%7D%29%22,%22hide%22:true,%22range%22:true,%22refId%22:%22C%22,%22interval%22:%22%22%7D,%7B%22refId%22:%22A%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%228Xt1pVoVk%22%7D,%22editorMode%22:%22builder%22,%22expr%22:%22sum%20by%28host%29%20%28rate%28subscriber_processed_batches%7Bhost%3D~%5C%22ams-bnc-val-00%7Cewr-bnc-val-00%5C%22,%20network%3D%5C%22benchmark%5C%22%7D%5B5m%5D%29%29%22,%22legendFormat%22:%22__auto%22,%22range%22:true,%22instant%22:true,%22hide%22:true%7D,%7B%22refId%22:%22B%22,%22datasource%22:%7B%22type%22:%22prometheus%22,%22uid%22:%228Xt1pVoVk%22%7D,%22editorMode%22:%22code%22,%22expr%22:%22sum%20by%20%28host%29%20%28rate%28sequencing_certificate_attempt%7Bhost%3D~%5C%22.%2A%5C%22,%20network%3D%5C%22benchmark%5C%22%7D%5B$__rate_interval%5D%29%29%22,%22legendFormat%22:%22__auto%22,%22range%22:true,%22instant%22:true%7D%5D,%22range%22:%7B%22from%22:%221678520418067%22,%22to%22:%221678552191281%22%7D%7D)

- 100 node geo distributed 
-- [**Before Batching** ~ 200-300 certs per second (2-3 rounds per second)](https://mysten.grafana.net/d/ORCQSHfVk/subscriber-bulk-fetch-dashboard?var-Environment=mysten-metrics-internal&var-network=benchmark&var-validator=atl-bnc-val-00&var-validator=atl-bnc-val-01&orgId=1&from=1679379442288&to=1679398925376&viewPanel=1)
-- [**After Batching** ~300-500 certs per second (3-5 rounds per second)](https://mysten.grafana.net/d/ORCQSHfVk/subscriber-bulk-fetch-dashboard?var-Environment=mysten-metrics-internal&var-network=benchmark&var-validator=atl-bnc-val-00&var-validator=atl-bnc-val-01&orgId=1&from=1679540599207&to=1679542836527)

### CODE REVIEW

File: narwhal/executor/src/metrics.rs [hunk 1]
```diff
@@ -38,6 +40,23 @@ pub struct ExecutorMetrics {
     /// Latency between the time when the batch has been
     /// created and when it has been fetched for execution
     pub batch_execution_latency: Histogram,
+    /// The number of batches per committed subdag to be fetched
+    pub committed_subdag_batch_count: Histogram,
+    /// Latency for time taken to fetch all batches for committed subdag
+    /// either from local or remote worker.
+    pub batch_fetch_for_committed_subdag_total_latency: Histogram,
+    /// The number of successful local batch fetches
+    pub successful_local_batch_fetch: IntCounter,
```
Comment (aschran) [inline narwhal/executor/src/metrics.rs:49]: Normally I see these sort of metrics done as one "counter" metric with a "status" field, might be simpler here?

So you'd have just a single metric, `batch_fetches` with labels `source`->local/remote, and `status`->success/failure/timeout.

File: narwhal/worker/src/handlers.rs [hunk 2]
```diff
@@ -28,6 +29,8 @@ use crate::TransactionValidator;
 #[path = "tests/handlers_tests.rs"]
 pub mod handlers_tests;
 
+pub const MAX_REQUEST_BATCHES_RESPONSE_SIZE: usize = 6_000_000;
```
Comment (aschran) [inline narwhal/worker/src/handlers.rs:32]: optional style suggestion, maybe put this inside the handler fn instead of at the top level of the file, so it's easier to find?

File: narwhal/worker/src/handlers.rs [hunk 3]
```diff
@@ -77,6 +80,30 @@ impl<V: TransactionValidator> WorkerToWorker for WorkerReceiverHandler<V> {
 
         Ok(anemo::Response::new(RequestBatchResponse { batch }))
     }
+
+    async fn request_batches(
+        &self,
+        request: anemo::Request<RequestBatchesRequest>,
+    ) -> Result<anemo::Response<RequestBatchesResponse>, anemo::rpc::Status> {
+        let batch_digests = request.into_body().batches;
```
Comment (aschran) [inline narwhal/worker/src/handlers.rs:88]: think it's worth adding a limit on `batch_digests.len()` in addition to size, to prevent doing excess work in case of a badly behaved peer? (esp because you fetch all of the requested batches regardless of how many are returned). worst-case, this could cause us to read from the DB 500KB*number of requested batches, right?

File: narwhal/executor/src/subscriber.rs [hunk 4]
```diff
@@ -231,8 +233,37 @@ impl<Network: SubscriberNetwork> Fetcher<Network> {
...
         };
 
+        let mut workers_for_certificates: HashSet<NetworkPublicKey> = HashSet::new();
+        let mut batch_digests: HashMap<WorkerId, HashSet<BatchDigest>> = HashMap::new();
+
+        for cert in &sub_dag.certificates {
+            for (digest, (worker_id, _)) in cert.header.payload.iter() {
+                let mut workers = self.network.workers_for_certificate(cert, worker_id);
+                workers.shuffle(&mut ThreadRng::default());
+                workers_for_certificates.extend(workers.clone());
```
Comment (aschran) [inline narwhal/executor/src/subscriber.rs:243]:
```suggestion
                let workers = self.network.workers_for_certificate(cert, worker_id);
                workers_for_certificates.extend(workers);
```

Does this work as well?

File: narwhal/executor/src/subscriber.rs [hunk 5]
```diff
@@ -243,148 +274,222 @@ impl<Network: SubscriberNetwork> Fetcher<Network> {
...
+
+            let _timer = self.metrics.subscriber_remote_fetch_latency.start_timer();
+            let mut stagger = Duration::from_secs(0);
+            let mut futures = vec![];
+            for worker in &workers {
+                let future = self.fetch_remote(stagger, worker.clone(), remaining_digests.clone());
+                futures.push(future.boxed());
+                // TODO: Make this a parameter, and also record workers / authorities that are down
+                // to request from them batches later.
+                stagger += Duration::from_millis(200);
```
Comment (aschran) [inline narwhal/executor/src/subscriber.rs:343]: wondering if this stagger is low enough that we will be stepping on our own feet for large requests?

for example if it normally takes us (say) 1 sec to download 6MB, then by the end even if the first worker we contact would be saturating our link, we'd be clogging it up with 4 other requests for the same thing that waste a lot of resources, slow down the original request, and then get thrown out?

File: narwhal/executor/src/subscriber.rs [hunk 6]
```diff
@@ -243,148 +274,222 @@ impl<Network: SubscriberNetwork> Fetcher<Network> {
...
+    ) -> HashMap<BatchDigest, Batch> {
         let _timer = self.metrics.subscriber_local_fetch_latency.start_timer();
         let worker = self.network.my_worker(&worker_id);
-        let payload = self.network.request_batch(digest, worker).await;
-        match payload {
-            Ok(Some(batch)) => {
-                debug!("Payload {} found locally", digest);
-                self.metrics.subscriber_local_hit.inc();
-                return Some(batch);
+        let request_batches_response = self.network.request_batches(digests, worker).await;
```
Comment (aschran) [inline narwhal/executor/src/subscriber.rs:380]: what if we have more than 6MB of batches available locally?

File: narwhal/executor/src/subscriber.rs [hunk 7]
```diff
@@ -243,148 +274,222 @@ impl<Network: SubscriberNetwork> Fetcher<Network> {
...
             attempt += 1;
             let deadline = Instant::now() + timeout;
             let request_batch_guard =
                 PendingGuard::make_inc(&self.metrics.pending_remote_request_batch);
-            let payload =
-                tokio::time::timeout_at(deadline, self.safe_request_batch(digest, worker.clone()))
-                    .await;
+            let response = tokio::time::timeout_at(
+                deadline,
+                self.safe_request_batches(remaining_digests.clone(), worker.clone()),
```
Comment (aschran) [inline narwhal/executor/src/subscriber.rs:432]: use anemo with_timeout on the rpc, rather than wrapping in tokio timeout?

this will require you to change your code in p2p.rs slightly to have the trait method take `request: impl anemo::types::request::IntoRequest<FetchBatchesRequest> + Send,`

example usage of with_timeout in narhwal:
https://github.com/MystenLabs/sui/blob/aad6425854d725a28ffee7d819b35913346ef212/narwhal/primary/src/synchronizer.rs#L773

File: narwhal/executor/src/subscriber.rs [hunk 8]
```diff
@@ -243,148 +274,222 @@ impl<Network: SubscriberNetwork> Fetcher<Network> {
...
+    ) -> anyhow::Result<HashMap<BatchDigest, Batch>> {
+        let fetched_batches = self
+            .network
+            .request_batches(digests.clone().into_iter().collect(), worker.clone())
+            .await?;
+        let mut verified_batches = HashMap::new();
+        for batch in fetched_batches {
+            let batch_digest = batch.digest();
+            if !digests.contains(&batch_digest) {
+                bail!("[Protocol violation] Worker {} returned batch with digest {} which is not part of the requested digests: {:?}", worker, batch_digest, digests );
```
Comment (aschran) [inline narwhal/executor/src/subscriber.rs:487]: possible to inline the format arguments?

**Review by aschran — COMMENTED** [review]
So excited for this PR! 🎉

File: narwhal/worker/src/handlers.rs (hunk 3 above)
Comment (arun-koshy) [reply narwhal/worker/src/handlers.rs]: Agreed, added a relatively high limit of 100 batches for now. Let me know if you think that limit is too high, need to get more data to make a more informed decision. But should protect us from any malicious peers.

File: narwhal/executor/src/subscriber.rs (hunk 6 above)
Comment (arun-koshy) [reply narwhal/executor/src/subscriber.rs]: Whoops forgot to a continuation token to `RequestBatches`. Thanks!

File: narwhal/executor/src/subscriber.rs (hunk 5 above)
Comment (arun-koshy) [reply narwhal/executor/src/subscriber.rs]: I agree with your concern, but we may also wait too long in between worker fetches if we go to a bad node first. I think because we are more likely to hit a good worker than a bad worker, I will increase this to **500ms*** and we can track how that performs. 

*assuming 10mbps is a fair minimum network expectation

File: narwhal/worker/src/handlers.rs [hunk 9]
```diff
@@ -77,6 +78,43 @@ impl<V: TransactionValidator> WorkerToWorker for WorkerReceiverHandler<V> {
...
         Ok(anemo::Response::new(RequestBatchResponse { batch }))
     }
+
+    async fn request_batches(
+        &self,
+        request: anemo::Request<RequestBatchesRequest>,
+    ) -> Result<anemo::Response<RequestBatchesResponse>, anemo::rpc::Status> {
+        const MAX_REQUEST_BATCHES_RESPONSE_SIZE: usize = 6_000_000;
+        const MAX_BATCH_DIGESTS: usize = 100;
+        let mut remaining_batch_digests = Vec::new();
```
Comment (aschran) [inline narwhal/worker/src/handlers.rs:88]: should `remaining_batch_digests` also include the digests that were truncated because they exceed `MAX_BATCH_DIGESTS`?

**Review by aschran — COMMENTED** [review]
ty for the updates :)

File: narwhal/worker/src/handlers.rs (hunk 3 above)
Comment (arun-koshy) [reply narwhal/worker/src/handlers.rs]: Turns out 100 was too low, updated it to 1_000

File: narwhal/worker/src/handlers.rs (hunk 9 above)
Comment (arun-koshy) [reply narwhal/worker/src/handlers.rs]: Yes! Included it as part of the remaining digests now

**Review by aschran — APPROVED** [review]

File: narwhal/executor/src/subscriber.rs [hunk 10]
```diff
@@ -243,151 +275,311 @@ impl<Network: SubscriberNetwork> Fetcher<Network> {
...
-            for (digest, (worker_id, _)) in cert.header.payload.iter() {
+            for (digest, (_, _)) in cert.header.payload.iter() {
                 self.metrics.subscriber_processed_batches.inc();
-
-                let mut workers = self.network.workers_for_certificate(cert, worker_id);
-
-                workers.shuffle(&mut ThreadRng::default());
+                let batch = fetched_batches
+                    .get(digest)
+                    .expect("[Protocol violation] Batch not found in fetched batches from certificate authors");
```
Comment (mwtian) [inline narwhal/executor/src/subscriber.rs:282]: Is the comment inside `expect()` accurate? I think the failure here actually means `fetch_batches_from_worker()` does not return all requested batches.

File: narwhal/executor/src/subscriber.rs [hunk 11]
```diff
@@ -231,8 +233,38 @@ impl<Network: SubscriberNetwork> Fetcher<Network> {
...
+                batch_digests,
+                workers_for_certificates.into_iter().collect(),
+            )
+            .await;
+        drop(fetched_batches_timer);
+
+        // Map all fetched batches to their respective certificates and submit as
+        // consensus output
+        for cert in &sub_dag.certificates {
+            let mut output_batches = Vec::with_capacity(num_batches);
```
Comment (mwtian) [inline narwhal/executor/src/subscriber.rs:267]: s/num_batches/cert.header.payload.len()

File: narwhal/executor/src/subscriber.rs [hunk 12]
```diff
@@ -231,8 +233,38 @@ impl<Network: SubscriberNetwork> Fetcher<Network> {
...
         };
 
+        let mut workers_for_certificates: HashSet<NetworkPublicKey> = HashSet::new();
+        let mut batch_digests: HashMap<WorkerId, HashSet<BatchDigest>> = HashMap::new();
+
+        let mut num_batches = 0;
         for cert in &sub_dag.certificates {
-            let mut batches = Vec::with_capacity(num_batches);
+            for (digest, (worker_id, _)) in cert.header.payload.iter() {
+                let workers = self.network.workers_for_certificate(cert, worker_id);
```
Comment (mwtian) [inline narwhal/executor/src/subscriber.rs:242]: I don't think we need to collect workers from the certificates. Instead, using all workers in the network should be better.

File: narwhal/executor/src/subscriber.rs [hunk 13]
```diff
@@ -243,151 +275,311 @@ impl<Network: SubscriberNetwork> Fetcher<Network> {
...
-            // TODO: Make this a parameter, and also record workers / authorities that are down
-            //       to request from them batches later.
-            stagger += Duration::from_millis(200);
+            // TODO: Can further parallelize this by worker_id if necessary.
+            // Only have one worker for now so will leave this for a future
+            // optimization.
+            let mut remaining_digests = digests.clone();
+
+            let local_batches = self
+                .try_fetch_locally(remaining_digests.clone().into_iter().collect(), worker_id)
```
Comment (mwtian) [inline narwhal/executor/src/subscriber.rs:320]: Maybe limit the size of the vec of remaining_digests here instead of limiting it in the handler?

File: narwhal/executor/src/subscriber.rs [hunk 14]
```diff
@@ -243,151 +275,311 @@ impl<Network: SubscriberNetwork> Fetcher<Network> {
...
+                        fetched_batches.insert(remote_batch_digest, remote_batch);
+                    }
+                }
+
+                if remaining_digests.is_empty() {
+                    return fetched_batches;
+                }
+            }
+
+            // Should never reach here, as we should keep looping until we have all batches remotely or locally.
```
Comment (mwtian) [inline narwhal/executor/src/subscriber.rs:372]: I think we will hit this log if there are more than 1 worker? I can't find the retry loop after exhausting all remote workers to fetch from.

File: narwhal/executor/src/subscriber.rs [hunk 15]
```diff
@@ -243,151 +275,311 @@ impl<Network: SubscriberNetwork> Fetcher<Network> {
...
+            }
+
+            if remaining_digests.is_empty() {
+                return fetched_batches;
+            }
+
+            let _timer = self.metrics.subscriber_remote_fetch_latency.start_timer();
+            let mut stagger = Duration::from_secs(0);
+            let mut futures = FuturesUnordered::new();
+            for worker in &workers {
```
Comment (mwtian) [inline narwhal/executor/src/subscriber.rs:343]: I believe we need to fetch from remote workers with the same worker index, but multi worker support may not be working so leaving a comment is fine ..

File: narwhal/executor/src/subscriber.rs [hunk 16]
```diff
@@ -243,151 +275,311 @@ impl<Network: SubscriberNetwork> Fetcher<Network> {
...
+                        "Batch {local_batch_digest:?} took {batch_fetch_duration} seconds to be fetched for execution since creation",
+                    );
+                    fetched_batches.insert(local_batch_digest, local_batch);
+                }
+            }
+
+            if remaining_digests.is_empty() {
+                return fetched_batches;
+            }
+
```
Comment (mwtian) [inline narwhal/executor/src/subscriber.rs:339]: Maybe have a loop to chunk remaining_digests per 1000 items?

Comment (arun-koshy) [reply narwhal/executor/src/subscriber.rs:349]: If I do chunk it here, would the idea be to issue those chunks in parallel or is this related to your idea of limiting outside of the handler? Wonder if it would saturate the network too much with so many potentially large responses being returned at the same time?

File: narwhal/worker/src/handlers.rs [hunk 17]
```diff
@@ -77,6 +78,40 @@ impl<V: TransactionValidator> WorkerToWorker for WorkerReceiverHandler<V> {
...
+
+    async fn request_batches(
+        &self,
+        request: anemo::Request<RequestBatchesRequest>,
+    ) -> Result<anemo::Response<RequestBatchesResponse>, anemo::rpc::Status> {
+        const MAX_REQUEST_BATCHES_RESPONSE_SIZE: usize = 6_000_000;
+        const MAX_REQUEST_BATCH_DIGESTS: usize = 1_000;
+
+        let mut digests_to_fetch = request.into_body().batch_digests;
+        let mut remaining_batch_digests = digests_to_fetch
```
[the remark this answers is not in the document]
Comment (mwtian) [reply narwhal/worker/src/handlers.rs:90]: `RequestBatches` is only issued from primary to worker, right? So there should be no risk of malicious overloading. But we do need rate limits and flow controls. My suggestions is to keep the size limit high (6MB is ok), and have the caller / primary determine number of digests. i.e. if batches to fetch should be chunked in the primary, and we only need a sanity check here.

We can come back and update this later.

[the remark this answers is not in the document]
Comment (arun-koshy) [reply narwhal/worker/src/handlers.rs:90]: But RequestBatches can go from primary to other workers so byzantine Primary A could overload Worker B and thus overload Primary B right?

File: narwhal/executor/src/subscriber.rs (hunk 12 above)
Comment (mwtian) [reply narwhal/executor/src/subscriber.rs:243]: Workers from the CommitttedSubdag may approximate all workers, so my preference is to keep things simple. But I don't mind keeping the current logic.

File: narwhal/executor/src/subscriber.rs (hunk 13 above)
Comment (mwtian) [reply narwhal/executor/src/subscriber.rs:330]: There is no bad requester to worry about since this is between primary and worker who trust each other.

File: narwhal/executor/src/subscriber.rs (hunk 16 above)
Comment (mwtian) [reply narwhal/executor/src/subscriber.rs:349]: I think chunking by # at the requestor side is cleaner. Limiting total size has to be done at the handler though. Since this is a trusted interface between primary and worker, I think we can divide the responsibilities?

File: narwhal/worker/src/handlers.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (mwtian) [reply narwhal/worker/src/handlers.rs:90]: Got it. In this case, I think we really should return an error if the number of digests received is over a threshold.

[the remark this answers is not in the document]
Comment (arun-koshy) [reply narwhal/worker/src/handlers.rs:90]: I can increase the limit to a number that is much larger than what we would expect and then return an error but for example even with 5k batch limit we do have nodes in normal operation with [P99 batch counts](https://mysten.grafana.net/d/ORCQSHfVk/subscriber-bulk-fetch-dashboard?var-Environment=mysten-metrics-internal&var-network=benchmark&var-validator=All&orgId=1&from=1679550342723&to=1679553942723&refresh=5s&viewPanel=5) for the `CommittedSubdag` close to 10k batches which I don't believe we should return an error for right?

[the remark this answers is not in the document]
Comment (aschran) [reply narwhal/worker/src/handlers.rs:90]: Just my arbitrary opinion, but I don't see why it's better to return an error? When there's already an interface to truncate responses based on size, why not do it based on count too? Seems simpler to keep the truncation logic all in one place rather than do it for size here and for count elsewhere.

[the remark this answers is not in the document]
Comment (aschran) [reply narwhal/worker/src/handlers.rs:90]: Alternatively, don't have a count limit at all and instead just have an internal limit within the worker on how many batches you will fetch from the DB at once? That seems like potentially best of both worlds.

[the remark this answers is not in the document]
Comment (mwtian) [reply narwhal/worker/src/handlers.rs:90]: My personal preference is for each end to utilize all their information so the protocol can be as efficient as possible. It does not seem to be an efficient protocol for the requestor to send well over 1000 batch digests in the request, and have everything after index 1000 sent back in the response. There are two areas I discussed with @arun-koshy 
1. It should be straightforward for the requestor to limit the number of batch digests in the request. In future we can also have parallel fetching but that is not needed for now.
2. The response contains a list of digests that exist on the worker but not included because of exceeding size limit. This is useful when fetching from own worker, but cannot be relied upon when fetching from other workers. We need to be careful when using this information.

**Review by mwtian — APPROVED** [review]
Overall looks good. Approve to unblock.

File: narwhal/worker/src/handlers.rs (hunk 17 above)
[the remark this answers is not in the document]
Comment (aschran) [reply narwhal/worker/src/handlers.rs:90]: > My personal preference is for each end to utilize all their information so the protocol can be as efficient as possible. It does not seem to be an efficient protocol for the requestor to send well over 1000 batch digests in the request, and have everything after index 1000 sent back in the response. There are two areas I discussed with @arun-koshy
> 
> 1. It should be straightforward for the requestor to limit the number of batch digests in the request. In future we can also have parallel fetching but that is not needed for now.
> 2. The response contains a list of digests that exist on the worker but not included because of exceeding size limit. This is useful when fetching from own worker, but cannot be relied upon when fetching from other workers. We need to be careful when using this information.

Fair points, Mingwei. I would recommend the following changes.
1. Entirely remove the batch count limit in RPC handler.
2. In its place, add an internal batch fetch limit. e.g. Fetch at most, say, ~200 batches at a time from DB. Add each fetched group to the response and terminate early if response size limit is exceeded. This limits RAM use exposure for malicious requester, while also allowing unlimited # of batches to be in one response in the more common case where they're all small enough to fit.
3. Remove the missing digest from response, the caller has to recalculate digests of what it got back anyway so it can decide on its own what was missing.

I think this addresses all three concerns?
- protocol efficiency
- batching as much as possible in each response, up to the limit
- protection against malicious peers
